neuralop/layers/discrete_continuous_convolution.py

```diff
@@ -0,0 +1,648 @@
+
+        if radius_cutoff <= 0.0:
+            raise ValueError("Error, radius_cutoff has to be positive.")
+
+        # compute how big the discrete kernel needs to be for the 2d convolution kernel to work
+        self.psi_local_h = math.floor(2*radius_cutoff * out_shape[0] / self.domain_length[0]) + 1
+        self.psi_local_w = math.floor(2*radius_cutoff * out_shape[1] / self.domain_length[1]) + 1
+
+        # compute the scale_factor
+        assert (in_shape[0] <= out_shape[0]) and (out_shape[0] % in_shape[0] == 0)
+        self.scale_h = out_shape[0] // in_shape[0]
+        assert (in_shape[1] <= out_shape[1]) and (out_shape[1] % in_shape[1] == 0)
+        self.scale_w = out_shape[1] // in_shape[1]
+
+        # psi_local is essentially the support of the hat functions evaluated locally
+        x = torch.linspace(-radius_cutoff, radius_cutoff, self.psi_local_h)
+        y = torch.linspace(-radius_cutoff, radius_cutoff, self.psi_local_w)
+        x, y = torch.meshgrid(x, y)
+        grid_in = torch.stack([x.reshape(-1), y.reshape(-1)])
+        grid_out = torch.Tensor([[0.0], [0.0]])
+
+        # compute quadrature weights on the incoming grid
+        self.q_weight = self.domain_length[0] * self.domain_length[1] / out_shape[0] / out_shape[1]
+        quad_weights = self.q_weight * torch.ones(self.psi_local_h * self.psi_local_w)
+
+        # precompute psi using conventional routines onto the local grid
+        # idx, vals = _precompute_convolution_tensor_2d(grid_out, grid_in, self.kernel_shape, quad_weights, radius_cutoff=radius_cutoff, periodic=False, normalize=False, transpose_normalization=True)
+        idx, vals = _precompute_convolution_tensor_2d(grid_in, grid_out, self.kernel_shape, quad_weights, radius_cutoff=radius_cutoff, periodic=False, normalize=True, transpose_normalization=False)
+
+        # extract the local psi as a dense representation
+        psi_loc = torch.zeros(self.kernel_size, self.psi_local_h*self.psi_local_w)
+        for ie in range(len(vals)):
+            f = idx[0, ie]; j = idx[2, ie]; v = vals[ie]
+            psi_loc[f, j] = v
+
+        # compute local version of the filter matrix
+        psi_loc = psi_loc.reshape(self.kernel_size, self.psi_local_h, self.psi_local_w)
+
+        self.register_buffer("psi_loc", psi_loc, persistent=False)
+
+    def get_psi(self):
+        """
+        Returns Psi
+        """
+
+        return self.psi_loc.permute(0, 2, 1).flip(dims=(-1,-2))
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Forward call. Expects an input of shape batch_size x in_channels x in_shape[0] x in_shape[1].
+        """
+
+        kernel = torch.einsum("kxy,ogk->ogxy", self.get_psi(), self.weight)
+
+        # padding is rounded down to give the right result when even kernels are applied
+        # Check https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html for output shape math
```

Comment:
Should this link be https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html instead?